### PR TITLE
Week 10a - Basic link/unlink provider login

### PR DESCRIPTION
**Pull request by connienguyen**
This PR allows users to now manage their connected logins if they are currently logged in. In terms of managing logins, this PR concerns only linking/unlinking a Google account.

+ Update UI flow (`ProfileViewController`, `SettingsTableViewController`, `LoginsManagerTableViewControllers`) to manage logins
+ Fix info window bug on map view
+ Update `LoginManager` so that all login management goes through Firebase
+ Migrate User model on local Realm database (delete old objects since primary key has changed)
+ Add unit test cases for linking a login to Firebase account

### CODE REVIEW

File: VOLA/VOLA/Services/LoginManager.swift [hunk 1]
```diff
@@ -38,27 +38,33 @@ final class LoginManager {
...
+        /// Log out of any providers
+        if let firebaseUser = FIRAuth.auth()?.currentUser {
+            for provider in firebaseUser.providerData {
+                if let loginProvider = LoginProvider(rawValue: provider.providerID) {
+                    switch loginProvider {
+                    case .facebook:
+                        FBSDKLoginManager().logOut()
+                    case .google:
+                        GIDSignIn.sharedInstance().signOut()
+                    default:
```
Comment (bphenriques) [inline VOLA/VOLA/Services/LoginManager.swift]: [Exaustive switch]

Switch cases should exhaustive when applicable. Reason: If anything changes the enum, the compiler will complain that that case was not handled

File: VOLA/VOLA/Services/LoginManager.swift [hunk 2]
```diff
@@ -38,27 +38,33 @@ final class LoginManager {
...
+                        break
+                    }
+                }
             }
 
             do {
-                try firebaseAuth.signOut()
+                try FIRAuth.auth()?.signOut()
```
Comment (bphenriques) [inline VOLA/VOLA/Services/LoginManager.swift]: [Unknown behaviour]

Is the `?` intentional? What if it is `nil`? What is the expected result?

In short, I dislike using the `?` operator as it is because it might make the code concise but then, it makes hard to track issues ;)

File: VOLA/VOLA/Services/LoginManager.swift [hunk 3]
```diff
@@ -77,4 +83,44 @@ final class LoginManager {
                 return true
             }
     }
+
+    /**
+    Link login to user's Firebase account
```
Comment (bphenriques) [inline VOLA/VOLA/Services/LoginManager.swift]: [Documentation]

Ditto:
Source: http://nshipster.com/swift-documentation/
Reference: https://github.com/systers/malaria-app-ios/blob/develop/Malaria%20Prevention%20App/Managers/MedicineHelpers/MedicineStats.swift

File: VOLA/VOLA/Services/LoginManager.swift [hunk 4]
```diff
@@ -77,4 +83,44 @@ final class LoginManager {
...
+     
+    - Returns: Boolean promise of whether or not connection was successful
+    */
+    func addConnectedLogin(_ strategy: AvailableConnectLoginStrategies) -> Promise<Bool> {
+        return strategy.connectLogin()
+    }
+
+    /**
+    Unlink a connected login from a user's Firebase account
+
```
Comment (bphenriques) [inline VOLA/VOLA/Services/LoginManager.swift]: [Documentation]

Ditto:
Source: http://nshipster.com/swift-documentation/
Reference: https://github.com/systers/malaria-app-ios/blob/develop/Malaria%20Prevention%20App/Managers/MedicineHelpers/MedicineStats.swift

File: VOLA/VOLA/Services/SocialNetworkingAuthenticationStrategy.swift [hunk 5]
```diff
@@ -15,6 +15,31 @@ protocol SocialNetworkingAuthenticationStrategy {
     func login() -> Promise<User>
 }
 
+// Extension on SocialNetworkingAuthenticationStrategy of helper functions that are common across strategies
```
Comment (bphenriques) [inline VOLA/VOLA/Services/SocialNetworkingAuthenticationStrategy.swift]: [English]

Extension of (...) with functions that are common among the different strategies

IMO, simpler to read

File: VOLA/VOLA/Services/SocialNetworkingAuthenticationStrategy.swift [hunk 6]
```diff
@@ -15,6 +15,31 @@ protocol SocialNetworkingAuthenticationStrategy {
...
+extension SocialNetworkingAuthenticationStrategy {
+    /**
+    Login to Firebase given a credential to a login provider
+     
+    - Parameters:
+        - credential: Credential for login provider
```
Comment (bphenriques) [inline VOLA/VOLA/Services/SocialNetworkingAuthenticationStrategy.swift]: ... for a ...

File: VOLA/VOLA/Services/SocialNetworkingAuthenticationStrategy.swift [hunk 7]
```diff
@@ -15,6 +15,31 @@ protocol SocialNetworkingAuthenticationStrategy {
     func login() -> Promise<User>
 }
 
+// Extension on SocialNetworkingAuthenticationStrategy of helper functions that are common across strategies
+extension SocialNetworkingAuthenticationStrategy {
+    /**
+    Login to Firebase given a credential to a login provider
```
Comment (bphenriques) [inline VOLA/VOLA/Services/SocialNetworkingAuthenticationStrategy.swift]: `credential to a login provider`? I would rephrase to Login to Firebase given a login provider

File: VOLA/VOLA/Services/SocialNetworkingAuthenticationStrategy.swift [hunk 8]
```diff
@@ -27,8 +52,8 @@ Available user authentication strategies
 enum AvailableLoginStrategies {
     case facebook
     case google(NSNotification)
-    case manualSignup(String, String, String)
-    case manualLogin(String, String)
+    case emailSignup(String, String, String)
```
Comment (bphenriques) [inline VOLA/VOLA/Services/SocialNetworkingAuthenticationStrategy.swift]: [enum case named types]

It's unclear what means the different `String`. Put labels on those :)

File: VOLA/VOLA/Services/SocialNetworkingAuthenticationStrategy.swift [hunk 9]
```diff
@@ -27,8 +52,8 @@ Available user authentication strategies
...
+    case emailLogin(String, String)
```
Comment (bphenriques) [inline VOLA/VOLA/Services/SocialNetworkingAuthenticationStrategy.swift]: ditto

File: VOLA/VOLA/Services/SocialNetworkingAuthenticationStrategy.swift (hunk 8 above)
Comment (bphenriques) [inline VOLA/VOLA/Services/SocialNetworkingAuthenticationStrategy.swift]: If you think that the fields are repeating, consider using a `struct` to wrap the `case` arguments

File: VOLA/VOLA/ViewControllers/LoginsManagerTableViewController.swift [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+
+        addNotificationObserver(NotificationName.googleDidSignIn, selector: #selector(googleDidSignIn(_:)))
+    }
+
+    override func viewWillDisappear(_ animated: Bool) {
+        super.viewWillDisappear(animated)
+
+        removeNotificationObserver(NotificationName.googleDidSignIn)
```
Comment (bphenriques) [inline VOLA/VOLA/ViewControllers/LoginsManagerTableViewController.swift]: [Correct unsubscribe]
The `removeNotificationObserver` should be on the `deinit` method as well ;)

Repeat this step on other parts of the code as required

File: VOLA/VOLA/ViewControllers/LoginsManagerTableViewController.swift [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+// MARK: - Table view delegate
+extension LoginsManagerTableViewController {
+    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        // TODO open connected login flow
+        let provider = viewModel.availableLogins[indexPath.row]
+        let isConnected = viewModel.loginIsConnected(provider)
+        let loginsCount = viewModel.connectedLoginsCount()
+        if isConnected && loginsCount > 1 {
```
Comment (bphenriques) [inline VOLA/VOLA/ViewControllers/LoginsManagerTableViewController.swift]: [unclear behaviour]

Document the `1` logic. Can't figure out its nature

File: VOLA/VOLA/ViewControllers/LoginsManagerTableViewController.swift [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+                break
+            }
+        }
+    }
+}
+
+// MARK: - NotificationObserver
+extension LoginsManagerTableViewController {
+    /**
+    Add connected login after user has successfully signed in with Google
```
Comment (bphenriques) [inline VOLA/VOLA/ViewControllers/LoginsManagerTableViewController.swift]: [Documentation]

Ditto:
Source: http://nshipster.com/swift-documentation/
Reference: https://github.com/systers/malaria-app-ios/blob/develop/Malaria%20Prevention%20App/Managers/MedicineHelpers/MedicineStats.swift

File: VOLA/VOLA/ViewModels/ConnectedLoginsViewModel.swift [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+
+import Foundation
+
+/// View model for display information about connected logins
+class ConnectedLoginViewModel {
+    let availableLogins = LoginProvider.allProviders
+
+    /**
+    Check whether user has connected login to their account
+
```
Comment (bphenriques) [inline VOLA/VOLA/ViewModels/ConnectedLoginsViewModel.swift]: [Documentation]

Ditto:
Source: http://nshipster.com/swift-documentation/
Reference: https://github.com/systers/malaria-app-ios/blob/develop/Malaria%20Prevention%20App/Managers/MedicineHelpers/MedicineStats.swift

File: VOLA/VOLA/ViewModels/ConnectedLoginsViewModel.swift [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+class ConnectedLoginViewModel {
+    let availableLogins = LoginProvider.allProviders
+
+    /**
+    Check whether user has connected login to their account
+     
+    - Parameters:
+        - provider: Login provider to check
+     
+    - Returns: Boolean of whether or not login is connected to Firebase account
```
Comment (bphenriques) [inline VOLA/VOLA/ViewModels/ConnectedLoginsViewModel.swift]: I think you can write with backticks the `Boolean`. It would preview as code on method preview on Xcode :)

File: VOLA/VOLA/ViewModels/SettingsViewModel.swift [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+            return UITableViewCell.self
+        case .connectedLogins:
+            return ManageLoginsCell.self
+        }
+    }
+}
+
+class SettingsViewModel {
+    static let loggedInSettings: [UserSetting] = [.theme, .connectedLogins]
+    static let anonUserSettings: [UserSetting] = [.theme]
```
Comment (bphenriques) [inline VOLA/VOLA/ViewModels/SettingsViewModel.swift]: [Unclear name]
`anon` ?

File: VOLA/VOLA/Views/Cells/ConnectedLoginCell.swift [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+class ConnectedLoginCell: UITableViewCell {
+    static let estimatedCellHeight: CGFloat = 75.0
+    
+    @IBOutlet weak var connectedLoginImage: CircleImageView!
+    @IBOutlet weak var connectedLoginStatus: TextLabel!
+
+    /// Configure the cell to show login provider connection status
+    func configureCell(_ provider: LoginProvider, connectedStatus: Bool) {
+        var labelText = provider.providerID
+        labelText += connectedStatus ? ": Connected" : ": Not Connected"
```
Comment (bphenriques) [inline VOLA/VOLA/Views/Cells/ConnectedLoginCell.swift]: [Localization]
Missing localization for those strings

File: VOLA/VOLA/Model/User.swift [hunk 17]
```diff
@@ -11,33 +11,44 @@ import RealmSwift
...
 }
 
 /// Model for User data
 class User: Object {
-
+    dynamic var uid: String = ""
     dynamic var name: String = ""
     dynamic var email: String = ""
-    dynamic var userTypeRaw: String = ""
+    dynamic var providersRawJoined: String = ""
```
Comment (connienguyen) [inline VOLA/VOLA/Model/User.swift]: okay, I'll rename, but it's the `rawValues` of `LoginProvider` since it's not a supported type by Realm.

Comment (bphenriques) [inline VOLA/VOLA/Model/User.swift]: Got it, understand. Then keep at it is. Is is possible to keep the `rawValue` private and add a public acessor that returns the enum value?

File: VOLA/VOLA/Services/ConnectLoginStrategy.swift [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+protocol ConnectLoginStrategy {
+    func connectLogin() -> Promise<Bool>
+}
+
+// Extension on ConnectLoginStrategy of helper functions that are common across strategies
+extension ConnectLoginStrategy {
+    /**
+    Connect authorized provider login to Firebase account given credentials
+     
+    - Parameters:
```
Comment (connienguyen) [inline VOLA/VOLA/Services/ConnectLoginStrategy.swift]: actually, from the link you provided, the `Parameters` and `Returns` notation method is also valid (i.e. can be viewed from XCode Quick Help).

But, I will update the documentation to be aligned correctly and address other issues 👍

File: VOLA/VOLA/Helpers/Util/RealmConfiguration.swift [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+//
+
+import Foundation
+import RealmSwift
+
+struct RealmConfiguration {
+    static let config = Realm.Configuration(
+        schemaVersion: 1,
+        migrationBlock: { migration, oldSchemaVersion in
+            if oldSchemaVersion < 1 {
```
Comment (connienguyen) [inline VOLA/VOLA/Helpers/Util/RealmConfiguration.swift]: I can remove it. I just left it in there in case you try to build from the previous version on your local repo in which case, you'll get an error.

Of course, in that case, you can always reset the simulator for dev purposes 🙂

File: VOLA/VOLA/Services/LoginManager.swift [hunk 20]
```diff
@@ -34,41 +34,47 @@ final class LoginManager {
...
+                        GIDSignIn.sharedInstance().signOut()
+                    case .email:
+                        break
+                    }
+                }
             }
 
             do {
-                try firebaseAuth.signOut()
+                if let auth = FIRAuth.auth() {
```
Comment (bphenriques) [inline VOLA/VOLA/Services/LoginManager.swift]: I did not explain well :) This is the same as you had before. The idea is that, if it is not expected then you should Log the unexpected behaviour. If it is expected, comment and maintain the `?` operator.

File: VOLA/VOLA/ViewControllers/LoginsManagerTableViewController.swift [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+    }
+}
+
+// MARK: - Table view delegate
+extension LoginsManagerTableViewController {
+    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        let provider = viewModel.availableLogins[indexPath.row]
+        let isConnected = viewModel.loginIsConnected(provider)
+        let loginsCount = viewModel.connectedLoginsCount()
+        if isConnected && loginsCount > 1 {
```
Comment (bphenriques) [inline VOLA/VOLA/ViewControllers/LoginsManagerTableViewController.swift]: I now understand. I would use something like

```swift

if loginsCount > 1 {
     removeConnectedLogin(...)
} else {
     signIn()    
}

...

func signIn() {
     guard !isConnected && loginsCount == 0 else {
            Logger.Error(...)
            return
      }

     switch provider() etc etc

}
```

Sorry for the lack of proper syntax. The point is to separate the logic onto different functions that improves code readability :)

File: VOLA/VOLA/Services/LoginManager.swift (hunk 20 above)
Comment (connienguyen) [inline VOLA/VOLA/Services/LoginManager.swift]: ah, okay. I will comment and maintain the `?` operator.

**Comment by bphenriques** [comment]
Added couple more comments

File: VOLA/VOLA/Helpers/Util/RealmConfiguration.swift (hunk 19 above)
Comment (psok) [inline VOLA/VOLA/Helpers/Util/RealmConfiguration.swift]: I agree with @bphenriques

File: VOLA/VOLA/Services/LoginManager.swift [hunk 22]
```diff
@@ -34,41 +34,46 @@ final class LoginManager {
...
+                        FBSDKLoginManager().logOut()
+                    case .google:
+                        GIDSignIn.sharedInstance().signOut()
+                    case .email:
+                        break
+                    }
+                }
             }
 
             do {
```
Comment (psok) [inline VOLA/VOLA/Services/LoginManager.swift]: Will it work if we put this block inside `case .email:`?

Comment (connienguyen) [inline VOLA/VOLA/Services/LoginManager.swift]: Firebase login is happening even if you use other methods of login, that way we can have one connected account. So, if you login with Facebook, you're actually authenticated your Facebook login, passing the credentials to Firebase, and logging into your Firebase connected account. That's why the Firebase login is outside email because it applies to all login methods.

File: VOLA/VOLA/Model/User.swift [hunk 23]
```diff
@@ -11,94 +11,78 @@ import RealmSwift
...
+    /// Provider identifier used by Firebase, primarily for unlinking a login
+    var providerID: String {
+        return rawValue
+    }
 }
 
 /// Model for User data
 class User: Object {
-
+    dynamic var uid: String = ""
```
Comment (connienguyen) [inline VOLA/VOLA/Model/User.swift]: This is generated by Firebase whether you login with email/password, google, or facebook. (Although facebook is not yet enabled).

File: VOLA/VOLA/Model/User.swift [hunk 24]
```diff
@@ -11,94 +11,78 @@ import RealmSwift
...
+    var providers: [LoginProvider?] {
+        return loginProvidersJoined.components(separatedBy: ",")
+            .map { LoginProvider(rawValue: $0) }
+            .filter { $0 != nil }
     }
     var imageURL: URL? {
         return URL(string: imageURLString)
     }
 
     /// Primary key for Realm object so that it can up updated in data store
```
Comment (connienguyen) [inline VOLA/VOLA/Model/User.swift]: sorry, I think I'm really tired from packing. What typo?

File: VOLA/VOLA/Helpers/Util/RealmConfiguration.swift (hunk 19 above)
Comment (connienguyen) [inline VOLA/VOLA/Helpers/Util/RealmConfiguration.swift]: okay, I'll remove it 👍

File: VOLA/VOLA/Model/User.swift (hunk 24 above)
Comment (bphenriques) [inline VOLA/VOLA/Model/User.swift]: "it can be updated" instead of "it can up updated"

File: VOLA/VOLA/Model/User.swift (hunk 23 above)
Comment (psok) [inline VOLA/VOLA/Model/User.swift]: Cool!

File: VOLA/VOLA/Services/LoginManager.swift (hunk 22 above)
Comment (psok) [inline VOLA/VOLA/Services/LoginManager.swift]: I see!

Comment (psok) [inline VOLA/VOLA/Services/LoginManager.swift]: Can you comment that out for future reference?

File: VOLA/VOLA/Helpers/Util/RealmConfiguration.swift (hunk 19 above)
Comment (shruti-gupta) [inline VOLA/VOLA/Helpers/Util/RealmConfiguration.swift]: agree with @bphenriques

File: VOLA/VOLA/Model/User.swift [hunk 25]
```diff
@@ -11,94 +11,78 @@ import RealmSwift
...
 
 /// Model for User data
 class User: Object {
-
+    dynamic var uid: String = ""
     dynamic var name: String = ""
     dynamic var email: String = ""
-    dynamic var userTypeRaw: String = ""
+    dynamic var loginProvidersJoined: String = ""
     dynamic var imageURLString: String = ""
```
Comment (shruti-gupta) [inline VOLA/VOLA/Model/User.swift]: [Naming]It is not generally a good practice to append data type in the variable name. Do we absolutely need string here? how about just imageURL?

Comment (connienguyen) [inline VOLA/VOLA/Model/User.swift]: `imageURL` is another property on this object that is a computed value from this. I have the `imageURLString` property in order for it to work with Realm, otherwise, we can't save the image url since the `URL` type is not supported by Realm.
